flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -140,6 +141,8 @@
 	/** Channels, which notified this input gate about available data. */
 	private final ArrayDeque<InputChannel> inputChannelsWithData = new ArrayDeque<>();
 
+	private final HashMap<Integer, InputChannel> channelsBlockedByCheckpoint = new HashMap<>();
```

Comment:
It is better to reuse the blocked states from `CheckpointBarrierAligner` if possible to avoid managing it duplicated in another place and causing potential inconsistency.